walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,79 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
+
+  // has one slot for every minute of a day
+  // The minute is true if all attendees are available in that minute, false
+  // otherwise
+  private boolean[] availability = new boolean[TimeRange.END_OF_DAY + 1];
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Arrays.fill(availability,true);
+
+    for (Event event : events) {
+      if (isEventRelevant(event, request.getAttendees())) {
+        recordBusyTime(event.getWhen());
+      }
+    }
+
+    Collection<TimeRange> availableTimes = new LinkedList<TimeRange>();
+
+    int startIndex = 0;
+
+    while (startIndex <= TimeRange.END_OF_DAY) {
+        if (!availability[startIndex]) {
+            startIndex ++;
+            continue;
+        }
+        
+        int endIndex = startIndex;
+
+        while (endIndex <= TimeRange.END_OF_DAY) {
+            if (!availability[endIndex]) {
+                break;
+            }
+            endIndex ++;
+        }
+
+
+        if (endIndex == TimeRange.END_OF_DAY && availability[endIndex]) {
```

Comment:
I've added a comment, but I'm not sure if it is clear to people that aren't me

Reply:
In the above loop:
while (endIndex <= TimeRange.END_OF_DAY) {
...
When endIndex == TimeRange.END_OF_DAY and availability[endIndex] is true, we should have already incremented the endIndex once more, making endIndex > TimeRange.END_OF_DAY. Is there any situation where this conditional is called?

Reply:
Oh good point, I didn't think of that. So that would mean that the structure of the while is achieving what I want it to right? And I can just remove that check completely?

Reply:
I suspect you could remove it entirely.

Here is where TDD would come in very handy; define a test that checks the desired behaviour, make the change, verify it with the test. In this case, you could have a test that should be able to verify the bounds of the final availability slot, which would be expected to succeed before and after the change. (Often with TDD you add a failing test, then make it pas).

Reply:
Ok, I'll make a test and see how it goes

Reply:
If some of the pre-existing tests cover this case, is it best practice to write a new test or to just use them?

Reply:
No need to duplicate test cases; generally if it's covered, then the behaviour is already being tested. Occasionally, it may be "incidentally" covered (in that some test data for one test just happens to cover this case, but that could change if the test set changed), in which case it can be useful to have a test that validates the specific thing you want to verify. I wouldn't worry about that too much here, but it's something to keep in mind for larger projects. Ideally; specific small tests that focus on a set of data to verify one part of the behaviour produces a more maintainable set of tests.

Reply:
I removed that section and the tests still pass